### PR TITLE
Intermediate expectation values.

### PR DESCRIPTION
**Pull request by 95-martin-orion**
Adds `simulate_moment_expectation_values` to the QSimSimulator API.

This method accepts a dict of (moment index: observable list), and calculates the requested observables at the specified moments. Special-case behavior is provided for calculating the same observable (or list of observables) at every moment.

### CODE REVIEW

**Comment by 95-martin-orion** [comment]
I have tested this locally, but have not attempted it on a GPU device.

File: qsimcirq/qsim_simulator.py [hunk 1]
```diff
@@ -718,3 +857,28 @@ def _translate_circuit(
             self._translated_circuits.append((circuit, translated_circuit))
 
         return translated_circuit
+
+    def _translate_moments(
```
Comment (sergeisakov) [inline qsimcirq/qsim_simulator.py:861]: There is no circuit memoization, right?

File: qsimcirq/qsim_simulator.py [hunk 2]
```diff
@@ -718,3 +857,28 @@ def _translate_circuit(
...
+        reindex = [0]
+        for moment in circuit:
+            subc = self._translate_circuit(
+                qsimc.QSimCircuit(cirq.Circuit(moment)),
+                translator_fn_name,
+                qubit_order,
+            )
+            reindex.append(len(subc.channels if is_noisy else subc.gates) + reindex[-1])
+
+        full_circuit = self._translate_circuit(circuit, translator_fn_name, qubit_order)
```
Comment (sergeisakov) [inline qsimcirq/qsim_simulator.py:883]: The circuit gets translated twice, right? The circuit translation is very slow. Can the code be changed in such a way that the circuit is translated only once (if this is not too much work)?

**Review by sergeisakov — CHANGES_REQUESTED** [review]
Good work!

File: pybind_interface/pybind_main.cpp [hunk 3]
```diff
@@ -658,6 +723,37 @@ class SimulatorHelper {
...
+      subncircuit.num_qubits = ncircuit.num_qubits;
+      subncircuit.channels = std::vector<Channel<Gate>>(
+        ncircuit.channels.begin() + begin,
+        ncircuit.channels.begin() + end
+      );
+
+      Simulator simulator = factory.CreateSimulator();
+      StateSpace state_space = factory.CreateStateSpace();
+
+      result = NoisyRunner::RunOnce(params, subncircuit, seed, state_space,
```
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply pybind_interface/pybind_main.cpp]: Done.

File: qsimcirq/qsim_simulator.py (hunk 1 above)
Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py:861]: Memoization occurs in `_translate_circuit`, which this calls for each moment. This has the added benefit that repeated moments will use their memoized translations. (Note however that this does not affect `CircuitOperation`s with repetitions, as a given `CircuitOperation` exists inside a single moment)

File: qsimcirq/qsim_simulator.py (hunk 2 above)
Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py:883]: Single translation should be possible - I'll need to add circuit-composition code on the C++ side, but I don't think that should be too complicated. I'll include this in this PR.

Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py]: Done: `compose[_noisy]_circuits` allows Python to merge two C++ circuits together.

**Comment by 95-martin-orion** [comment]
The tests pass locally for me, which suggests some version-specific misbehavior. I'll check it out.

**Comment by sergeisakov** [comment]
> The tests pass locally for me, which suggests some version-specific misbehavior. I'll check it out.

The tests also pass locally for me.

File: qsimcirq/qsim_simulator.py [hunk 4]
```diff
@@ -718,3 +857,33 @@ def _translate_circuit(
...
+        is_noisy: bool,
+    ):
+        """Translates a circuit by moments.
+
+        Returns:
+            A tuple containing the translated circuit and the gate indices
+            corresponding to moment boundaries.
+        """
+        reindex = [0]
+        full_circuit = qsim.NoisyCircuit() if is_noisy else qsim.Circuit()
```
Comment (sergeisakov) [inline qsimcirq/qsim_simulator.py:875]: I would probably do this in a different way. I would modify `translate_cirq_to_qsim` and  `translate_cirq_to_qtrajectory` in qsim_circuit.py to return the gate indices corresponding to moment boundaries in addition to returning translated circuits. In this case, `qsim.compose_noisy_circuits` and `qsim.compose_circuits` are not required.

Comment (95-martin-orion) [reply qsimcirq/qsim_simulator.py]: Sounds good to me, let's see how the CI tests feel about it.

**Comment by 95-martin-orion** [comment]
Okay, now I'm suspicious of issues external to this PR. The sum total of C++ changes between the latest commit and the last passing commit is a single line - and it's a comment. Meanwhile about half of the tests are now failing and at least one of them segfaults...

**Review by sergeisakov — APPROVED** [review]
LGTM

File: pybind_interface/pybind_main.cpp [hunk 5]
```diff
@@ -559,6 +559,70 @@ class SimulatorHelper {
...
+        results[i] = helper.get_expectation_value(counts);
+        begin = end;
+      }
+      return results;
+    }
+
+    // Aggregate expectation values for noisy circuits.
+    for (unsigned i = 0; i < opsums_and_qubit_counts.size(); ++i) {
+      auto& counts = std::get<1>(opsums_and_qubit_counts[i]);
+      results[i] = std::vector<std::complex<double>>(counts.size(), 0);
```
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply pybind_interface/pybind_main.cpp]: Done.

**Pull request merged**